File: car/list_ports.py
#!/usr/bin/env python3
"""List serial ports with stable /dev/serial/by-id paths when available."""

from pathlib import Path

from serial.tools import list_ports

CH343_VID = 0x1A86
BY_ID_DIR = Path("/dev/serial/by-id")
# Always present on the Pi with GPS unplugged. A new line is the GPS module.
IGNORE = {"ttyS0", "ttyAMA0", "serial0"}


def by_id_for(device: str) -> str | None:
    if not BY_ID_DIR.is_dir():
        return None
    try:
        target = Path(device).resolve()
    except OSError:
        return None
    for link in sorted(BY_ID_DIR.iterdir()):
        try:
            if link.resolve() == target:
                return str(link)
        except OSError:
            continue
    return None


for p in list_ports.comports():
    vid = p.vid or 0
    name = Path(p.device).name
    hay = f"{p.description or ''} {p.manufacturer or ''}".upper()
    if vid == CH343_VID or "CH343" in hay or "CH340" in hay or "WCH" in hay:
        tag = "  <-- LoRa"
    elif name.startswith("rfcomm") or name in IGNORE or p.device == "/dev/obd":
        tag = ""
    else:
        tag = "  <-- GPS"
    stable = by_id_for(p.device)
    primary = stable or p.device
    bits = [
        primary,
        f"({p.device})" if stable else None,
        f"vid={vid:04X}" if p.vid else None,
        p.manufacturer,
        p.description,
    ]
    print("  ".join(x for x in bits if x) + tag)

File: base_station/list-ports.js
#!/usr/bin/env node
/**
 * List USB-TO-LoRa serial ports for the base station laptop.
 * Windows → COMx via list-ports-windows.js
 * Linux/macOS → /dev/serial/by-id/... when available
 *
 * Pi car ports: python3 ../car/list_ports.py
 */
import fs from "fs";
import path from "path";
import { SerialPort } from "serialport";

if (process.platform === "win32") {
  await import("./list-ports-windows.js");
  process.exit(0);
}

const CH343_VID = 0x1a86;
const BY_ID_DIR = "/dev/serial/by-id";

function byIdFor(devicePath) {
  if (!devicePath || !fs.existsSync(BY_ID_DIR)) return null;
  let target;
  try {
    target = fs.realpathSync(devicePath);
  } catch {
    return null;
  }
  for (const name of fs.readdirSync(BY_ID_DIR).sort()) {
    const link = path.join(BY_ID_DIR, name);
    try {
      if (fs.realpathSync(link) === target) return link;
    } catch {
      /* skip */
    }
  }
  return null;
}

const ports = await SerialPort.list();
if (!ports.length) {
  console.log("No serial ports found.");
  process.exit(0);
}
for (const p of ports) {
  const vid = p.vendorId ? parseInt(p.vendorId, 16) : 0;
  const hay = `${p.manufacturer || ""} ${p.friendlyName || ""} ${p.path || ""}`.toUpperCase();
  const tag =
    vid === CH343_VID || hay.includes("CH343") || hay.includes("CH340") || hay.includes("WCH")
      ? "  <-- LoRa?"
      : "";
  const stable = byIdFor(p.path);
  console.log(
    [
      stable || p.path,
      stable ? `(${p.path})` : null,
      p.vendorId ? `vid=${p.vendorId}` : null,
      p.productId ? `pid=${p.productId}` : null,
      p.manufacturer || null,
      p.friendlyName || p.serialNumber || null,
    ]
      .filter(Boolean)
      .join("  ") + tag
  );
}
